### PR TITLE
Configure CORS

### LINKED ISSUE #1553: No CORS problems when using cnd during the workshop/hackathon
## Problem

A browser-based application will currently run into CORS limitations as it tries to connect to cnd because the HTTP API is running on a different port than the webserver of the application.

## Goal

Allow browser-based applications to interact with cnd.

## Recommendation

Add an optional configuration parameter `cors` to the config file in the `http_api` section.

Possible values should be :

- `all`: allow requests from all hosts
- `none` (default value): don't include cors headers
- `host:port`: only allow requests from the given host and port

Create-comit-app should then set this value to `all` for the instances it launches during developlment.

This recommendation will require to implement our own `cors` module for warp because the existing one cannot handle the `all` case. Fortunately, the cors headers support wildcards (`*`), so we just have to write a little module that sets all the headers to `*` if `all` is specified.

### PR DESCRIPTION
**Pull request by luckysori**
Fixes #1553.

### CODE REVIEW

**Comment by D4nte** [comment]
If it's ugly (and it is) then implement a custom deserializer.

File: cnd/src/http_api/route_factory.rs [hunk 1]
```diff
@@ -33,14 +34,29 @@ pub fn create<
         + SaveRfc003Messages,
 >(
     peer_id: PeerId,
+    allowed_foreign_origins: AllowedForeignOrigins,
     dependencies: D,
+    peer_id: PeerId,
```
Comment (tcharding) [inline cnd/src/http_api/route_factory.rs]: We don't need the second `peer_id` here :).

File: cnd/src/main.rs [hunk 2]
```diff
@@ -138,9 +138,16 @@ fn spawn_warp_instance<
     runtime: &mut tokio::runtime::Runtime,
     dependencies: D,
 ) {
-    let routes = route_factory::create(peer_id, dependencies);
+    let routes = route_factory::create(
+        dependencies,
+        settings.http_api.cors.allowed_foreign_origins.clone(),
+        peer_id,
```
Comment (tcharding) [inline cnd/src/main.rs]:
```
    let routes = route_factory::create(
        peer_id,
        dependencies,
        settings.http_api.cors.allowed_foreign_origins.clone(),
```

File: cnd/src/config/file.rs [hunk 3]
```diff
@@ -252,14 +286,52 @@ mod tests {
...
+    fn cors_deserializes_correctly() {
+        let file_contents = vec![
+            r#"
+            allowed_origins = "all"
+            "#,
+            r#"
+             allowed_origins = "none"
+            "#,
+            r#"
+             allowed_origins = ["http://localhost:8000", "https://192.168.1.55:3000"]
```
Comment (D4nte) [inline cnd/src/config/file.rs]: 👍

**Pull request merged**